### PR TITLE
handle serialization/deserialization for prob fx dataframes

### LINKED ISSUE #495: TypeError: Only pandas Series with a localized DatetimeIndex is supported
serialization fails for processed probabilistic forecast timeseries because they're stored as a DataFrame

Sentry Issue: [WORKER-API-R](https://sentry.io/organizations/solararbiter-rx/issues/1748464895/?referrer=github_integration)

```
TypeError: Only pandas Series with a localized DatetimeIndex is supported
(3 additional frame(s) were not displayed)
...
  File "solarforecastarbiter/reports/main.py", line 278, in compute_report
    )(report.report_id, raw_report)
  File "solarforecastarbiter/reports/main.py", line 225, in wrapper
    out = f(*args, **kwargs)
  File "solarforecastarbiter/io/api.py", line 1088, in post_raw_report
    report_id, raw_report)
  File "solarforecastarbiter/io/api.py", line 1010, in post_raw_report_processed_data
    fxobs.forecast_values)}
  File "solarforecastarbiter/io/utils.py", line 298, in serialize_timeseries
    'Only pandas Series with a localized DatetimeIndex is supported')
```

### PR DESCRIPTION
**Pull request by wholmgren**
- [x] Closes #495 .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

So far I have a test that reproduces the failure.

@alorenzo175 I started trying to make `serialize_timeseries` work with DataFrames and it's turning out to be less straightforward than I thought it would be. The `schema` has `'column': 'value'` - do we want to change this to `'columns': [{df.columns}]` and then `deserialize_timeseries` has some conditional on this? Or do we want to let the calling functions serialize the DataFrame column by column, and the reverse for deserialize?

I can keep hacking at this but feel free to take over if you have a good idea of what to do.

### CODE REVIEW

**Comment by alorenzo175** [comment]
> The schema has 'column': 'value' - do we want to change this to 'columns': [{df.columns}] and then deserialize_timeseries has some conditional on this?

I think that is ok. maybe worth adding another key to schema to specify output should be dataframe or series then process column as appropriate. Definitely keep some data serialized with the current schema in the tests to make sure deserialize works with old and new

**Comment by wholmgren** [comment]
@alorenzo175 can you check this approach before I add the tests? Example below

<details>

```python
index = pd.date_range(start='20200101', periods=2, freq='12h', tz='UTC')
s = pd.Series([1., 2], index=index)
df = pd.DataFrame({0.: [0., 1], 0.5: [0.5, 1.5]}, index=index)

s_str = utils.serialize_timeseries(s)
s_str

'{"schema":{"version": 0, "orient": "records", "timezone": "UTC", "column": "value", "index": "timestamp", "dtype": "float64", "objtype": "Series"},"data":[{"timestamp":"2020-01-01T00:00:00Z","value":1.0},{"timestamp":"2020-01-01T12:00:00Z","value":2.0}]}'

df_str = utils.serialize_timeseries(df)
df_str

'{"schema":{"version": 0, "orient": "records", "timezone": "UTC", "column": [0.0, 0.5], "index": "timestamp", "dtype": ["float64", "float64"], "objtype": "DataFrame"},"data":[{"timestamp":"2020-01-01T00:00:00Z","0.0":0.0,"0.5":0.5},{"timestamp":"2020-01-01T12:00:00Z","0.0":1.0,"0.5":1.5}]}'

utils.deserialize_timeseries(s_str)
timestamp
2020-01-01 00:00:00+00:00    1.0
2020-01-01 12:00:00+00:00    2.0
Name: value, dtype: float64

utils.deserialize_timeseries(df_str)

  | 0.0 | 0.5
-- | -- | --

0.0 | 0.5
1.0 | 1.5
```

</details>

**Comment by alorenzo175** [comment]
A few things:
- probably different enough to warrang version 1
- does the column selection work? I would think there might be issues since the column has to be a string key in the records, but maybe it went in as a float, so a roundtrip should have float column labels.

I think the column keys should be string no matter and avoid precision issues, but that might require changes elsewhere

**Comment by wholmgren** [comment]
> does the column selection work? I would think there might be issues since the column has to be a string key in the records, but maybe it went in as a float, so a roundtrip should have float column labels.

Creating a dataframe with string columns and then serializing leads to this json: `"column": ["0.", "0.5"]`

Whereas creating a dataframe with float columns and then serializing leads to this json: `"column": [0.0, 0.5]`

`pd.read_json` creates string column names, perhaps due to the presence of 'timestamp' or maybe that's just how it always works. In any case, the `str_col_map = {str(col): col for col in schema['column']}` does the job of ensuring the roundtrip works. 

And column selection works in either case so long as you use the right type.

> I think the column keys should be string no matter and avoid precision issues, but that might require changes elsewhere

`api.get_probabilistic_forecast_values` returns a DataFrame with float columns because it creates a dict with keys `cv.constant_value` as it loops over the forecasts. Adding `str` is easy there, though I guess I'm not sure if it's really the right thing to do or not.

`calculator._transform_prob_forecast_value_and_prob` performs indexing using the column as specified, and contains a couple lines with `float(col)` so a string would be covered here.

We might need to slightly change a couple of tests. 

I think a couple of tests would need to be updated too.

File: solarforecastarbiter/io/tests/test_api.py [hunk 1]
```diff
@@ -588,7 +584,7 @@ def test_apisession_get_prob_forecast_values(
     for col in col_values:
         new_cv = copy.deepcopy(cvs_json)
         new_cv['constant_value'] = col
-        new_cv['forecast_id'] = 'CV' + str(int(col))
+        new_cv['forecast_id'] = 'CV' + str(int(float(col)))
```
Comment (alorenzo175) [inline solarforecastarbiter/io/tests/test_api.py:587]: odd to bother with float to then take int, but ok

File: solarforecastarbiter/io/tests/test_io_utils.py [hunk 2]
```diff
@@ -216,7 +229,9 @@ def test_adjust_timeseries_for_interval_label_series(label, exp):
         pytest.mark.xfail(strict=True, type=TypeError)]),
     pytest.param(pd.Series([], dtype=float, index=pd.DatetimeIndex([])),
                  marks=[pytest.mark.xfail(strict=True, type=TypeError)]),
-    pytest.param(TEST_DATA, marks=[
+    TEST_DATAFRAME,
+    EMPTY_TIME_DATAFRAME,
```
Comment (alorenzo175) [inline solarforecastarbiter/io/tests/test_io_utils.py:233]: Can you test the serialized output of this explicitly? I'm a little wary about assuming keys are present in the schema for this case

**Comment by wholmgren** [comment]
Thanks. Go ahead and merge if you’d like or I can in an hour  or so.

**Pull request merged**